### PR TITLE
fix(ci): refresh firebase token and check valid to start test for specific build

### PR DESCRIPTION
**Pull request by milapjoshi**
Signed-off-by: milap joshi <milap.joshi@gmail.com>

## Test Plan
https://magma-ci.web.app/

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by quentinDERORY — APPROVED** [review]
LGTM  but can we resolve/hide lint issues?

**Comment by talkhasib** [comment]
Please rebase as some files have changed upstream

File: ci-scripts/test_agent/Tester.py [hunk 1]
```diff
@@ -32,6 +32,7 @@ def test_ended(self):
         # Get test results and prepare report
         verdict, report = self.load_test_results(dbfile="/tmp/test_res_pickle")
         # Callback
+        print("Test report created pushing to DB")
```
Comment (talkhasib) [inline ci-scripts/test_agent/Tester.py]: Tester has no awareness of DB. Maybe: "Test ended. Invoking callback function"

File: ci-scripts/test_agent/Tester.py [hunk 2]
```diff
@@ -59,27 +60,36 @@ def start_test(self, workload, build, test_done_callback):
...
             proc.wait()
             call_ended()
             return
 
-        thread = threading.Thread(
-            target=run_hil_thread, args=(self.test_ended, ["./run_test.sh"])
-        )
-        thread.start()
+        if self.current_build.val()["agw"]["valid"]:
+            magma_build = self.current_build.val()["agw"]["artifacts"]["downloadUri"]
```
Comment (talkhasib) [inline ci-scripts/test_agent/Tester.py]: If you can switch to using build['agw']['packages'] that would be great. This way we can quickly deprecate the old field
See example from latest builds on https://magma-ci.web.app/

File: ci-scripts/test_agent/main.py [hunk 3]
```diff
@@ -64,8 +65,10 @@ def main():
                     {"state": tester.get_state(), "current_workload": None}
                 )
 
+        time.sleep(1)
+        db_client.user = db_client.auth.refresh(db_client.user["refreshToken"])
         db_client.update_worker_state(num_of_testers, testers_state)
-        time.sleep(15)
+        time.sleep(14)
```
Comment (talkhasib) [inline ci-scripts/test_agent/main.py]: why 14?

File: ci-scripts/test_agent/main.py [hunk 4]
```diff
@@ -64,8 +65,10 @@ def main():
                     {"state": tester.get_state(), "current_workload": None}
                 )
 
+        time.sleep(1)
+        db_client.user = db_client.auth.refresh(db_client.user["refreshToken"])
```
Comment (talkhasib) [inline ci-scripts/test_agent/main.py]: Tokens are good for 1 hr. Refreshing every 15 seconds is too much.
Lets make it refresh every 15-30 min

File: ci-scripts/test_agent/main.py [hunk 5]
```diff
@@ -28,6 +28,7 @@ class WorkerState:
 
 
 def test_done_callback(tester_id, workload, verdict, report):
+    db_client.user = db_client.auth.refresh(db_client.user["refreshToken"])
```
Comment (talkhasib) [inline ci-scripts/test_agent/main.py]: No need for this, if we refresh on  main loop

**Review by talkhasib — CHANGES_REQUESTED** [review]

File: ci-scripts/test_agent/main.py (hunk 4 above)
Comment (milapjoshi) [reply ci-scripts/test_agent/main.py]: 👍

File: ci-scripts/test_agent/FirebaseClient.py [hunk 6]
```diff
@@ -34,14 +34,16 @@ def __init__(self):
...
             "timestamp": int(time.time()),
             "verdict": verdict,
             "workload": workload.val(),
             "report": report,
         }
-        self.db.child("workers").child(self.config["agent_id"]).child("reports").push(
-            data, self.user["idToken"]
+        # Use build_id as key for report. Overwrites any existing report
+        self.db.child("workers").child(self.config["agent_id"]).child("reports").child(build_id).set(
+            data, self.user["idToken"],
```
Comment (talkhasib) [inline ci-scripts/test_agent/FirebaseClient.py]: Can you rebase. Master already has this.

**Comment by talkhasib** [comment]
please resolve all the pep8 errors

File: ci-scripts/test_agent/FirebaseClient.py (hunk 6 above)
Comment (milapjoshi) [reply ci-scripts/test_agent/FirebaseClient.py]: i did and just pushed. this should be resolved now

**Comment by milapjoshi** [comment]
> please resolve all the pep8 errors

@talkhasib - most pep8 errors are resolved. There are two with using subprocess module itself. if warning must be resolved. I will have to devise separate way to integrate HIL with agent. I believe subprocess module usage should not be concern.

File: ci-scripts/test_agent/main.py [hunk 7]
```diff
@@ -33,9 +35,18 @@ def test_done_callback(tester_id, workload, verdict, report):
...
+
+    print("Refreshing Firebase Token:", datetime.now())
+    db_client.user = db_client.auth.refresh(db_client.user["refreshToken"])
+    threading.Timer(RefreshTokenTime, do_token_refresh).start()
+
+
 def main():
     state = WorkerState.READY
     testers = []
+    do_token_refresh()
```
Comment (talkhasib) [inline ci-scripts/test_agent/main.py]: schedule it instead
```threading.Timer(RefreshTokenTime, do_token_refresh).start()```

File: ci-scripts/test_agent/main.py [hunk 8]
```diff
@@ -33,9 +35,18 @@ def test_done_callback(tester_id, workload, verdict, report):
     return
 
 
+def do_token_refresh():
+    """Run in its own thread and refresh token every 30 min."""
+
+    print("Refreshing Firebase Token:", datetime.now())
+    db_client.user = db_client.auth.refresh(db_client.user["refreshToken"])
```
Comment (talkhasib) [inline ci-scripts/test_agent/main.py]: move this to a function inside the client. Avoid accessing members of the client here

File: ci-scripts/test_agent/Tester.py [hunk 9]
```diff
@@ -59,26 +60,51 @@ def start_test(self, workload, build, test_done_callback):
...
+                print(f"test {self.id} started on workload")
+            else:
+                print("No Magma Package found in packages list")
+                self.callback(self.id, self.current_workload, "fail", "NA")
+                self.current_workload = None
+                self.current_build = None
+                self.state = TesterState.READY
+
+        else:
+            print("Build check is not valid")
```
Comment (talkhasib) [inline ci-scripts/test_agent/Tester.py]: "Invalid AGW build {id}. Terminate test"

File: ci-scripts/test_agent/Tester.py [hunk 10]
```diff
@@ -59,26 +60,51 @@ def start_test(self, workload, build, test_done_callback):
...
+                thread = threading.Thread(
+                    target=run_hil_thread,
+                    args=(self.test_ended, ["./run_test.sh " + magma_build]),
+                )
+                thread.start()
+
+                self.state = TesterState.BUSY
+                print(f"test {self.id} started on workload")
+            else:
+                print("No Magma Package found in packages list")
```
Comment (talkhasib) [inline ci-scripts/test_agent/Tester.py]: s/Package/package

**Review by talkhasib — COMMENTED** [review]
few nits

**Comment by talkhasib** [comment]
Best if you can make the client handle the token refresh transparently to the user (main.py)

File: ci-scripts/test_agent/FirebaseClient.py [hunk 11]
```diff
@@ -12,10 +12,11 @@
 """
 
 import json
-import time
-
 import pyrebase
+import threading
+import time
 
+from datetime import datetime
```
Comment (talkhasib) [inline ci-scripts/test_agent/FirebaseClient.py:17]: run "isort [filename>]" locally to get this fixed

File: ci-scripts/test_agent/FirebaseClient.py [hunk 12]
```diff
@@ -32,6 +33,8 @@ def __init__(self):
         )
         # Get a reference to the database service
         self.db = self.firebase.database()
+        # token refresh timer
+        self.RefreshTokenTime = 1800
```
Comment (talkhasib) [inline ci-scripts/test_agent/FirebaseClient.py:38]: Sorry if I was not clear. I suggest scheduling the token refresh here and not in main.py
main.py does not need to be aware of this. Unless this causes a multi threading race issue. does it?

File: ci-scripts/test_agent/main.py [hunk 13]
```diff
@@ -36,6 +36,7 @@ def test_done_callback(tester_id, workload, verdict, report):
 def main():
     state = WorkerState.READY
     testers = []
+    threading.Timer(db_client.RefreshTokenTime, db_client.do_token_refresh).start()
```
Comment (talkhasib) [inline ci-scripts/test_agent/main.py]: see comment above

File: ci-scripts/test_agent/FirebaseClient.py (hunk 12 above)
Comment (milapjoshi) [reply ci-scripts/test_agent/FirebaseClient.py:38]: oh i see, sorry for back and forth. i think i cover that now.

**Review by talkhasib — APPROVED** [review]
Thanks @milapjoshi, LGTM

**Pull request merged**